### PR TITLE
Add fast path for match checking

### PR DESCRIPTION
**Pull request by ishitatsuyuki**
This adds a fast path that would reduce the complexity to linear on matches consisting of only variant patterns (i.e. enum matches). (Also see: #7462) Unfortunately, I was too lazy to add a similar fast path for constants (mostly for integer matches), ideally that could be added another day.

TBH, I'm not confident with the performance claims due to the fact that enums tends to be small and FxHashMap could add a lot of overhead.

r? @Mark-Simulacrum 

needs perf

### CODE REVIEW

**Comment by ishitatsuyuki** [comment]
Seems that all perf results are nearly identical. I guess that match checking wasn't dominating the compile time then.

It seems that this PR only helps in pathological cases like that generated by the script below. Some crate in the ecosystem  might be doing the same kind of code generation, though, and this can be become useful in such cases.

Python script:
```python
print("""#![allow(warnings)]
enum A {""")
for i in range(8192):
    print(f"    Var{i},")
print("""}
fn main() {
    match A::Var0 {""")
for i in range(8192):
    print(f"        A::Var{i} => {{}}")
print("""    }
}""")
```

Old rustc takes around 2.7s, this build takes around 1.7s.

**Comment by panstromek** [comment]
Btw. Addressing your concerns about FxHashMap overhead - https://github.com/rust-lang/rust/pull/72412 added a MiniMap for similar concerns (it's SSO map backed by ArrayVec). You can try using that here, seems like the map will always hold small number of elements, too, if I understand it correctly.

**Comment by ishitatsuyuki** [comment]
It seems that `MiniMap` is pretty limited - it doesn't support iteration and `entry`. So for now I'm not going to make the switch.

File: compiler/rustc_mir_build/src/thir/pattern/_match.rs [hunk 1]
```diff
@@ -522,18 +536,65 @@ impl<'p, 'tcx> Matrix<'p, 'tcx> {
                 self.push(row)
             }
         } else {
-            self.0.push(row);
+            if self.patterns.is_empty() {
```
Comment (oli-obk) [inline compiler/rustc_mir_build/src/thir/pattern/_match.rs:539]: Please put this code into a separate function and document it in much more detail. This module is already super complex and adding a fast path to it that does not change behaviour does not help with that complexity.

**Comment by ishitatsuyuki** [comment]
I added some more comments explaining the code per request.

**Comment by oli-obk** [comment]
Thanks, that helps a lot!

One thing that we could do is to add your pathological case to the perf test suite, then, once that's in, this PR will show an improvement on the perf bot and we thus guard against future regressions.

Another thing is to `debug_assert!` that the result of the fast path is the same as the slow path. This will require running both paths if debug assertions are activated, but I think it could be reasonably added. What do you think?

**Comment by ishitatsuyuki** [comment]
I added the debug assertions, it has a clone that is not really necessary but I thought the code would be easier to maintain this way.

As for perf, I have mixed feelings since the case I mentioned above is only one kind of match that is slow; I think we could add it once we have support for integer cases like https://github.com/rust-lang/rust/issues/7462#issuecomment-299335065.

**Comment by oli-obk** [comment]
You can add perf tests for both the integer case and your case within one test, then we are at least tracking them, and any changes to it will show up. If the tests just take a few seconds, that's perfect for perf, as it won't really slow down the perf test suite

**Comment by oli-obk** [comment]
Implementation wise this PR lgtm now, so we could merge it and add the perf test later if you prefer. We'd lose the visible improvement in the perf tests, but I'm not sure how much value that has.

**Comment by ishitatsuyuki** [comment]
Perf PR is up at https://github.com/rust-lang/rustc-perf/pull/769.

**Comment by ishitatsuyuki** [comment]
The `match-stress-enum` perf benchmark is now fully deployed.

**Pull request merged**